lib/bolt/analytics.rb

```diff
@@ -0,0 +1,143 @@
+        NoopClient.new
+      else
+        unless config.key?('user-id')
+          config['user-id'] = SecureRandom.uuid
+          write_config(config_file, config)
+        end
+
+        Client.new(config['user-id'])
+      end
+    rescue StandardError => e
+      logger.debug "Failed to initialize analytics client, analytics will be disabled: #{e}"
+      NoopClient.new
+    end
+
+    def self.load_config(filename)
+      if File.exist?(filename)
+        YAML.load_file(filename)
+      else
+        {}
+      end
+    end
+
+    def self.write_config(filename, config)
+      FileUtils.mkdir_p(File.dirname(filename))
+      File.write(filename, config.to_yaml)
+    end
+
+    class Client
+      attr_reader :user_id
+
+      def initialize(user_id)
+        @logger = Logging.logger[self]
+        @http = HTTPClient.new
+        @user_id = user_id
+        @executor = Concurrent.global_io_executor
+      end
+
+      def screen_view(screen)
+        screen_view_params = {
+          # Type
+          t: 'screenview',
+          # Screen Name
+          cd: screen
+        }
+
+        submit(base_params.merge(screen_view_params))
+      end
+
+      def event(category, action)
+        event_params = {
+          # Type
+          t: 'event',
+          # Event Category
+          ec: category,
+          # Event Action
+          ea: action
+        }
+
+        submit(base_params.merge(event_params))
+      end
+
+      def submit(params)
+        # Handle analytics submission in the background to avoid blocking the
+        # app or polluting the log with errors
+        Concurrent::Future.execute(executor: @executor) do
+          @logger.debug "Submitting analytics: #{JSON.pretty_generate(params)}"
+          @http.post(TrackingURL, query: params)
+          @logger.debug "Completed analytics submission"
+        end
+      end
+
+      # These parameters have terrible names. See this page for complete documentation:
+      # https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters
+      def base_params
+        {
+          v: PROTOCOL_VERSION,
+          # Client ID
+          cid: @user_id,
+          # Tracking ID
+          tid: TRACKING_ID,
+          # Application Name
+          an: APPLICATION_NAME,
+          # Application Version
+          av: Bolt::VERSION,
+          # Anonymize IPs
+          aip: true,
+          # User locale
+          ul: Locale.current.to_rfc
+        }
+      end
+
+      # If the user is running a very fast command, there may not be time for
+      # analytics submission to complete before the command is finished. In
+      # that case, we give a little buffer for any stragglers to finish up.
+      # 250ms strikes a balance between accomodating slower networks while not
+      # introducing a noticeable "hang".
+      def finish
+        @executor.shutdown
+        @executor.wait_for_termination(0.25)
```

Comment:
I could see spinning off another process to handle submissions so we can shut Bolt down immediately, but that's a far more complicated architecture that's probably not warranted yet.

Reply:
Right now we only make a single request, at the very start, so this is only an issue if you're doing something *very* fast. I can only make it cause a problem if I'm running something over the `local` transport.

In the future, if we're submitting events, they might happen toward the end, but they'll also reuse the existing HTTP connection and should be *significantly* faster.